### PR TITLE
Fix Backward compatibility with 2.0.0 Spark

### PR DESCRIPTION
**Pull request by hbhanawat**
## Changes proposed in this pull request

Snappy 0.7 should work with Spark versions >= 2.0.0 and versions < = 2.0.2 in connector mode. This PR fixes the issues related to that. 
- Changed the spark version for snappy-core to 2.0.0. 
- Copied three classes from Spark to Snappy as there is no compatibility between 2.0.0 and 2.0.2 versions of those classes. RunCommand, DescribeTableCommand, ExecutedCommandExec. 
- Added reflection to create JacksonGenerator object as its apply method has changed between two versions. 
- Added reflection to find out the parameter 'child' for CreateTableUsingAsSelect. 

## Patch testing
 
Clean precheckin.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/SnappySession.scala [hunk 1]
```diff
@@ -1529,6 +1530,8 @@ object SnappySession extends Logging {
     val (cachedRDD, shuffleDeps, rddId, localCollect) = executedPlan match {
       case _: ExecutedCommandExec =>
         throw new EntryExistsException("uncached plan", df) // don't cache
+      case _: SnappyExecutedCommandExec =>
+        throw new EntryExistsException("uncached plan", df) // don't cache
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/SnappySession.scala]: Can combine with ExecuteCommandExec case above with an '|'

File: core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+ * **Copy of DescribeTableCommand to ensure a compatibility between
+ * Spark 2.0.0 and Snappy Spark 2.0.2
+ *
+ * Command that looks like
+ * {{{
+ *   DESCRIBE [EXTENDED|FORMATTED] table_name partitionSpec?;
+ * }}}
+ */
+case class SnappyDescribeTableCommand(
+    table: TableIdentifier,
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala]: Can we avoid "Snappy" prefix for each of the new plan types? Say:

SnappyRunnableCommand => ExecuteCommand
SnappyDescribeTableCommand => DescribeTable

File: core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    }
+  }
+
+  private def describeSchema(schema: StructType, buffer: ArrayBuffer[Row]): Unit = {
+    schema.foreach { column =>
+      val comment =
+        if (column.metadata.contains("comment")) column.metadata.getString("comment") else null
+      append(buffer, column.name, column.dataType.simpleString, comment)
+    }
+  }
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala]: Please file a JIRA to enable showing extended properties for SnappyData tables.

File: core/src/main/scala/org/apache/spark/sql/backwardcomp/sparkPlans.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package org.apache.spark.sql.backwardcomp
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.execution.SparkPlan
+
+private[sql] case class SnappyExecutedCommandExec(cmd: SnappyRunnableCommand) extends SparkPlan {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/backwardcomp/sparkPlans.scala]: Too much Snappy in names. Say: SnappyExecutedCommandExec => ExecutedCommand

File: core/src/main/scala/org/apache/spark/sql/sources/StoreStrategy.scala [hunk 5]
```diff
@@ -32,39 +32,52 @@ object StoreStrategy extends Strategy {
...
+      import scala.reflect.runtime.{universe => ru}
+      import ru._
+      val rm = runtimeMirror(this.getClass.getClassLoader)
+      val pr = rm.reflect(a)
+      // 2.0.0 parameter name: child
+      // 2.0.2 parameter name: query
+      // Use reflection to find the value of parameter
+      val childterm = typeOf[CreateTableUsingAsSelect].members.find(a =>
+        a.fullName.contains("query") || a.fullName.contains("child"))
+      val value = pr.reflectField(childterm.get.asTerm).get
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/sources/StoreStrategy.scala]: Instead of reflection a cleaner way would be using Product.productElement (see the patch I had sent):
```scala
    case c@CreateTableUsingAsSelect(tableIdent, provider, partitionCols,
    _, mode, opts, _) =>
      val query = c.productElement(6).asInstanceOf[LogicalPlan]
      // CreateTableUsingSelect is only invoked by DataFrameWriter etc
      // so that should support both builtin and external tables
      ExecuteCommand(CreateMetastoreTableUsingSelect(tableIdent, None,
        None, None, SnappyContext.getProvider(provider, onlyBuiltIn = false),
        temporary = false, partitionCols, mode, opts, query,
        isBuiltIn = false)) :: Nil
```

File: core/src/main/scala/org/apache/spark/sql/sources/StoreStrategy.scala [hunk 6]
```diff
@@ -32,39 +32,52 @@ object StoreStrategy extends Strategy {
...
+      SnappyExecutedCommandExec(drop) :: Nil
 
     case DMLExternalTable(name, storeRelation: LogicalRelation, insertCommand) =>
-      ExecutedCommandExec(ExternalTableDMLCmd(storeRelation, insertCommand)) :: Nil
+      SnappyExecutedCommandExec(ExternalTableDMLCmd(storeRelation, insertCommand)) :: Nil
 
     case PutIntoTable(l@LogicalRelation(t: RowPutRelation, _, _), query) =>
-      ExecutedCommandExec(PutIntoDataSource(l, t, query)) :: Nil
+      SnappyExecutedCommandExec(PutIntoDataSource(l, t, query)) :: Nil
+    case r: SnappyDescribeTableCommand => SnappyExecutedCommandExec(r) :: Nil
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/sources/StoreStrategy.scala]: redundant

File: core/src/test/scala/org/apache/spark/sql/store/SnappyCatalogSuite.scala [hunk 7]
```diff
@@ -413,7 +413,10 @@ abstract class CatalogTestUtils {
         CatalogColumn("a", "int"),
         CatalogColumn("b", "string")),
       partitionColumnNames = Seq("a", "b"),
-      bucketColumnNames = Seq("col1"))
+      bucketColumnNames = Seq("col1"),
+      // Pass the properties so that the hack applied in SnappyExternalCatalog.createTable
+      // takes affect
+      properties = Map("spark.sql.sources.provider" -> "column"))
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/SnappyCatalogSuite.scala]: Why is this required now? What used to happen with 2.0.2 without this?

Comment (hbhanawat) [inline core/src/test/scala/org/apache/spark/sql/store/SnappyCatalogSuite.scala]: Well, I was trying to fix the failures in SnappyCatalogSuite (both with 2.0.0 and 2.0.2) and thought it could be the reason. But it isn't it. But kept it here as it takes it to the correct block which should be tested.

File: core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala (hunk 3 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala]: SNAP-1294

File: core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala (hunk 2 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/sql/backwardcomp/logicalPlans.scala]: I find Snappy prefix better as it conveys the class from which it was copied immediately. But, I believe it's a matter a perception. Going ahead with what you have suggested.
